benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/time_step_mod.f90

```diff
@@ -69,860 +71,808 @@ SUBROUTINE invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, &
+                sshn_t_device, ssha_t_device)
+            ierr = clEnqueueNDRangeKernel(cmd_queues(1), kernel_field_copy_code3, &
+                2, C_NULL_PTR, C_LOC(globalsize), C_LOC(localsize), 0, &
+                C_NULL_PTR, C_NULL_PTR)
+        else
+            swap = ssha_t%device_ptr
+            ssha_t%device_ptr = sshn_t%device_ptr
+            sshn_t%device_ptr = swap
+
+            swap = ua%device_ptr
+            ua%device_ptr = un%device_ptr
+            un%device_ptr = swap
+
+            swap = va%device_ptr
+            va%device_ptr = vn%device_ptr
+            vn%device_ptr = swap
+        endif
+
         ! Block until all kernels have finished
         ierr = clFinish(cmd_queues(1))
+        CALL check_status('Barrier end of iteration', ierr)
 
-        ! Specify device data retrieving methods
-        ssha_t%read_from_device_f => read_fortcl
-        sshn_t%read_from_device_f => read_fortcl
-        sshn_u%read_from_device_f => read_fortcl
-        sshn_v%read_from_device_f => read_fortcl
-        un%read_from_device_f => read_fortcl
-        vn%read_from_device_f => read_fortcl
-        ua%read_from_device_f => read_fortcl
-        va%read_from_device_f => read_fortcl
-
-
+        if (first_time) then
+            first_time = .false.
+        endif
     END SUBROUTINE invoke_time_step
 
-    subroutine read_fortcl(from, to, nx, ny, width)
-        use FortCL, only: read_buffer
-        use iso_c_binding, only: c_intptr_t, c_int
-        integer(c_intptr_t), intent(in) :: from
-        real(go_wp), dimension(:,:), intent(inout) :: to
-        integer, intent(in) :: nx, ny, width
-        ! Use width instead of nx in case there is padding elements
-        call read_buffer(from, to, int(width*ny, kind=8))
-    end subroutine read_fortcl
-
-    SUBROUTINE continuity_code_set_args(kernel_obj, ssha_t, sshn_t, &
-            sshn_u, sshn_v, hu, hv, un, vn, area_t, rdt)
-        USE clfortran, ONLY: clSetKernelArg
-        USE iso_c_binding, ONLY: c_sizeof, c_loc, c_intptr_t
+    subroutine initialise_device_buffer(field)
+        USE fortcl, ONLY: create_rw_buffer
+        type(r2d_field), intent(inout), target :: field
+        integer(kind=c_size_t) size_in_bytes
+        IF (.NOT. field%data_on_device) THEN
+            size_in_bytes = int(field%grid%nx*field%grid%ny, 8) * &
+                            c_sizeof(field%data(1,1))
+            ! Create buffer on device
+            field%device_ptr = transfer(create_rw_buffer(size_in_bytes), &
+                                         field%device_ptr)
+            field%data_on_device = .true.
+            field%read_from_device_f => read_opencl
+            field%write_to_device_f => write_opencl
+        END IF
+    end subroutine initialise_device_buffer
+
+    ! This OpenCL manual implementation only ever read/writes whole buffers
+    ! at once, so the coarse-grain (full rows) read/write functions below
+    ! already provide sufficient performance.
```

Comment:
Please also mention that `blocking` is ignored.

Reply:
Done